src/fifelse.c

```diff
@@ -201,105 +201,104 @@ SEXP fifelseR(SEXP l, SEXP a, SEXP b, SEXP na) {
-  SEXPTYPE type0 = TYPEOF(value0);
-  int64_t len0=xlength(cons0), len2=len0;
-  if (isS4(value0) && !INHERITS(value0, char_nanotime)) {
-    error(_("S4 class objects (except nanotime) are not supported. Please see https://github.com/Rdatatable/data.table/issues/4131."));
-    // otherwise 'invalid type/length (S4/1) in vector allocation' from test 2132.3
-  }
-  SEXP ans = PROTECT(allocVector(type0, len0));
-  SEXP tracker = PROTECT(allocVector(INTSXP, len0));
-  int *restrict p = INTEGER(tracker);
-  copyMostAttrib(value0, ans);
-
-  bool nonna=!isNull(na);
-  if (nonna) {
-    if (xlength(na) != 1) {
-      error(_("Length of 'default' must be 1."));
-    }
-    SEXPTYPE tn = TYPEOF(na);
-    if (tn==LGLSXP && LOGICAL(na)[0]==NA_LOGICAL) {
-      nonna = false;
-    } else {
-      if (tn != type0) {
-        error(_("Resulting value is of type %s but 'default' is of type %s. "
-                "Please make sure that both arguments have the same type."), type2char(type0), type2char(tn));
-      }
-      if (!R_compute_identical(PROTECT(getAttrib(value0,R_ClassSymbol)),  PROTECT(getAttrib(na,R_ClassSymbol)), 0)) {
-        error(_("Resulting value has different class than 'default'. "
-                "Please make sure that both arguments have the same class."));
-      }
-      UNPROTECT(2);
-      if (isFactor(value0)) {
-        if (!R_compute_identical(PROTECT(getAttrib(value0,R_LevelsSymbol)),  PROTECT(getAttrib(na,R_LevelsSymbol)), 0)) {
-          error(_("Resulting value and 'default' are both type factor but their levels are different."));
-        }
-        UNPROTECT(2);
-      }
-    }
-  }
-
+  int nprotect = 0, l = 0;
+  int64_t len0=0, len1=0, len2=0, idx=0;
+  SEXP ans = R_NilValue, value0 = R_NilValue, tracker = R_NilValue, whens = R_NilValue, thens = R_NilValue;
+  PROTECT_INDEX Iwhens, Ithens;
+  PROTECT_WITH_INDEX(whens, &Iwhens); nprotect++;
+  PROTECT_WITH_INDEX(thens, &Ithens); nprotect++;
+  SEXPTYPE type0;
+  // naout means if the output is scalar logic na
+  bool imask = true, naout = false, idefault = false;
+  int *restrict p = NULL;
   const int n = narg/2;
   for (int i=0; i<n; ++i) {
-    SEXP cons = PROTECT(i==0 ? cons0 : eval(SEXPPTR_RO(args)[2*i], rho)); // protect cons0 again for easy unprotect at the end of this loop
-    SEXP outs = PROTECT(i==0 ? value0 : eval(SEXPPTR_RO(args)[2*i+1], rho));
-    if (isS4(outs) && !INHERITS(outs, char_nanotime)) {
+    idefault = i == (n - 1); // mark if the current eval is the `default` on R side
+    REPROTECT(whens = eval(SEXPPTR_RO(args)[2*i], rho), Iwhens);
+    REPROTECT(thens = eval(SEXPPTR_RO(args)[2*i+1], rho), Ithens);
+    if (isS4(thens) && !INHERITS(thens, char_nanotime)) {
       error(_("S4 class objects (except nanotime) are not supported. Please see https://github.com/Rdatatable/data.table/issues/4131."));
     }
-    if (!isLogical(cons)) {
+    if (!isLogical(whens)) {
```

Comment:
We should say what type we encountered, rather than simply it was not type logical (so user can more easily identify the error).

Reply:
Agreed, added the typeof for the arg